libdnf/sack/query.cpp

```diff
@@ -769,6 +831,116 @@ Query::addFilter(HyNevra nevra, bool icase)
     return 0;
 }
 
+void
+Query::Impl::filterNevraStrict(int cmpType, const char **matches)
+{
+    Pool *pool = dnf_sack_get_pool(sack);
+    std::vector<NevraID> compareSet;
+    const unsigned nmatches = g_strv_length((gchar**)matches);
+    compareSet.reserve(nmatches);
+    for (unsigned int i = 0; i < nmatches; ++i) {
+        const char * nevraPattern = matches[i];
+        if (!nevraPattern)
+            throw std::runtime_error("Query can not accept NULL for STR match");
+        size_t nevraPatternLength =  strlen(nevraPattern);
+        const char * evrStart = NULL;
+        const char * arch = NULL;
+        bool evrDeliminator = false;
+        bool epochDeliminator = false;
+        size_t possition;
+        for (possition = nevraPatternLength-2; possition != 0; --possition) {
```

Comment:
this function looks awful.

```c
          char *arch = strrchr (nevra, '.');
          *arch++ = '\0';
          char *version = strrchr (nevra, '-');
          char *release = version;
          *release = '\0';
          version = strrchr (nevra, '-');
          *release = '-';
          *version++ = '\0';
          Id nid = pool_str2id (pool, nevra, 1);
          Id evrid = pool_str2id (pool, version, 1);
          Id aid = pool_str2id (pool, arch, 1);
```

This is everything you need

Reply:
@ignatenkobrain 
Your code seems simple but:
1. remove of epoch 0 is missing (if it is present)
2. `pool_str2id (pool, nevra, 1);` adds searched string into pool if it not exist in pool, so use 0 argument instead 1
3. multiple usage of strrchr() is not optimal - each call means go forward and found the end of the string and then go backward. But original code is not optimal too.

Reply:
I just wanted to give a hint how to implement thing in much easier way.

Reply:
Easier and much more readable way.

Reply:
I modified it, but implementation is different than proposal.